core/src/main/java/org/apache/iceberg/schema/UnionByNameVisitor.java

```diff
@@ -180,6 +180,17 @@ private void updateColumn(Types.NestedField field, Types.NestedField existingFie
     }
   }
 
+  private boolean compatibleType(Type newType, Type existingType) {
+    if (existingType.isPrimitiveType()) {
+      // Primitive -> Wider Primitive
+      return newType.isPrimitiveType()
+          && TypeUtil.isPromotionAllowed(newType, existingType.asPrimitiveType());
```

Comment:
shouldn't this be `TypeUtil.isPromotionAllowed(existingType, newType.asPrimitiveType())`? The signature of the method is `public static boolean isPromotionAllowed(Type from, Type.PrimitiveType to)`

Reply:
just reading https://github.com/apache/iceberg/pull/11419#issuecomment-2471220679 now. I think it's a bit confusing with the method naming and how `isPromotionAllowed` is called to a reader of the code. I think I'd agree with @RussellSpitzer's comment about naming this more like `ignoreTypeChange` or something along those lines. Also the comments should probably indicate that `primitive type -> narrower primitive type` is effectively allowed

Reply:
This is all my fault since I wrote those names in my suggestion in the first place, sorry for churn

Reply:
Fixed it last update